content/performance/preallocate-arrays.md

```diff
@@ -0,0 +1,41 @@
+---
+title: "Preallocate Arrays"
+section: "Performance"
+menu:
+  toc:
+    parent: "performance"
+    weight: 20
+---
+
+## Problem
+
+Your code is performance sensitive and suffers from poor performance due to arrays being resized while in use.
+
+## Solution
+
+It's common for new Pony programmers to create arrays that use the default allocation:
+
+```pony
+// allocate space for no entries
+let small: Array[U8] = Array[U8]
+```
+
+Instead, preallocate more space than you expect to use:
+
+```pony
+// preallocate space for at least 2048 entries
+let not_so_small: Array[U8] = Array[U8](2048)
+```
+
+## Discussion
+
+Arrays are an incredibly flexible data structure. Programmers can deploy them in situations where the amount of data needed will vary over time. If the array is too small and more space is needed to hold additional items, the array will be expanded so that it can hold more items.
```

Comment:
The concept of "too small" is very context-dependent. I think this discussion would benefit from a bit more detail on some of the factors that might go into determining how big the preallocation should be. It would also be useful to discuss the tradeoffs between smaller and larger preallocations.

Reply:
I don't think it is:

 "too small and more space is needed to hold additional items".

is a complete statement not just "too small".

How about "isn't large enough to hold all items without expanding it's size".

Reply:
I think you could at least include a bit more discussion. In my mind there are two important cases:

1. You know the number of items that will be added to the array. In that case, preallocate for improved performance.
2. You know some range of the number of items that will be added to the array. In that case, you might want to preallocate some number of items. If you preallocate too many then you'll have a bunch of unused space that's wasted. If you don't then you'll incur a penalty every time the array needs to be resized. There's a tradeoff here that makes sense to folks who have done this kind of work and understand what's going on, but it may not be clear to people who are new to this, so I think it should be addressed.